include/boost/beast/core/buffers_generator.hpp

```diff
@@ -0,0 +1,229 @@
+#include <boost/asio/async_result.hpp>
+#include <boost/asio/compose.hpp>
+#include <boost/asio/coroutine.hpp>
+
+#include <boost/beast/core/buffer_traits.hpp>
+#include <boost/beast/core/stream_traits.hpp>
+#include <boost/beast/core/detail/type_traits.hpp>
+
+#include <boost/throw_exception.hpp>
+#include <type_traits>
+
+namespace boost {
+namespace beast {
+
+/** Determine if type satisfies the <em>BuffersGenerator</em> requirements.
+
+    This metafunction is used to determine if the specified type meets the
+    requirements for a buffers generator.
+
+    The static member `value` will evaluate to `true` if so, `false` otherwise.
+
+    @tparam T a type to check
+*/
+#ifdef BOOST_BEAST_DOXYGEN
+template <class T>
+struct is_buffers_generator
+    : integral_constant<bool, automatically_determined>
+{
+};
+#else
+template<class T, class = void>
+struct is_buffers_generator
+    : std::false_type
+{
+};
+
+template<class T>
+struct is_buffers_generator<
+    T, detail::void_t<decltype(
+        typename T::const_buffers_type(
+            std::declval<T&>().prepare(
+                std::declval<error_code&>())),
+        std::declval<T&>().consume(
+            std::size_t{})
+    )>> : std::true_type
+{
+};
+#endif
+
+//----------------------------------------------------------
+
+namespace detail {
+
+template <
+    class AsyncWriteStream,
+    class BuffersGenerator>
+struct write_buffers_generator_op
+    : boost::asio::coroutine
+{
+    write_buffers_generator_op(AsyncWriteStream& stream,
+                               BuffersGenerator generator)
+        : stream_(stream)
+        , generator_(std::move(generator))
+    {
+    }
+
+    template <class Self>
+    void operator()(Self& self,
+                    error_code ec = {},
+                    std::size_t n = 0)
+    {
+        BOOST_ASIO_CORO_REENTER(*this)
+        {
+            for(;;)
+            {
+                BOOST_ASIO_CORO_YIELD
+                {
+                    auto cb = generator_.prepare(ec);
+                    if (ec)
+                        goto complete;
+                    if (boost::beast::buffer_bytes(cb) == 0)
+                        goto complete;
+                    stream_.async_write_some(
+                        cb, std::move(self));
+                }
+                generator_.consume(n);
+
+                total_ += n;
+            }
+
+        complete:
+            self.complete(ec, total_);
+        }
+    }
+
+private:
+    AsyncWriteStream& stream_;
+    BuffersGenerator generator_;
+    std::size_t total_ = 0;
+};
+
+} // detail
+
+//----------------------------------------------------------
+
+template <
+    class SyncWriteStream,
+    class BuffersGenerator
+#if ! BOOST_BEAST_DOXYGEN
+    ,
+    typename std::enable_if<
+        is_buffers_generator<typename std::decay<
+            BuffersGenerator>::type>::value>::type* =
+        nullptr
+#endif
+    >
+size_t
+write(SyncWriteStream& stream,
```

Comment:
This needs to go in `<boost/beast/core/impl/buffers_generator.hpp>` so we don't force the doc toolchain to parse it needlessly.

Reply:
Just checked the brand new javadoc. Trying to find out whether I can @ref link concepts (probably not?)

Reply:
You can add _BuffersGenerator_ as a concept to the docs and to the docca config:
https://github.com/boostorg/beast/blob/9d23bec2bc523223a377582302bd28d04cb514ed/doc/xsl/custom-overrides.xsl#L18

Reply:
Oh shiiiiiiiny. What do you say I create a separate issue to improve the docs here, because there's some loose ends (we could use a "Generator oriented streaming operations" section/page, with that table you alluded to earlier).